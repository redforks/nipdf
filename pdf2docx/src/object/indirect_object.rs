use super::{Object, ObjectId};

#[derive(Debug, Clone, PartialEq)]
pub struct IndirectObject<'a>(ObjectId, Object<'a>);

impl<'a> IndirectObject<'a> {
    pub fn new(id: u32, generation: u16, object: Object<'a>) -> Self {
        Self(ObjectId::new(id, generation), object)
    }

    pub fn id(&self) -> ObjectId {
        self.0
    }

    pub fn object(&self) -> &Object<'a> {
        &self.1
    }

    pub fn take(self) -> Object<'a> {
        self.1
    }
}
