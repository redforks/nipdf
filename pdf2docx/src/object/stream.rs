use std::{
    borrow::{Borrow, Cow},
    fmt::Display,
    iter::repeat,
    str::from_utf8,
};

use bitstream_io::{BigEndian, BitReader};
use image::{DynamicImage, GrayImage, Luma, RgbImage};
use log::error;
use once_cell::unsync::Lazy;
use pdf2docx_macro::pdf_object;

use crate::{
    ccitt::Flags,
    file::ObjectResolver,
    parser::{ws_prefixed, ParseResult},
};

use super::{Dictionary, Name, Object, ObjectValueError, SchemaDict};

const KEY_FILTER: &[u8] = b"Filter";
const KEY_FILTER_PARAMS: &[u8] = b"DecodeParms";
const KEY_FFILTER: &[u8] = b"FFilter";

const FILTER_FLATE_DECODE: &str = "FlateDecode";
const FILTER_CCITT_FAX: &str = "CCITTFaxDecode";
const FILTER_DCT_DECODE: &str = "DCTDecode";
const FILTER_ASCII85_DECODE: &str = "ASCII85Decode";
const FILTER_RUN_LENGTH_DECODE: &str = "RunLengthDecode";
const FILTER_JPX_DECODE: &str = "JPXDecode";

#[cfg(test)]
const B_FILTER_FLATE_DECODE: &[u8] = FILTER_FLATE_DECODE.as_bytes();
#[cfg(test)]
#[allow(unused)]
const B_FILTER_CCITT_FAX: &[u8] = FILTER_CCITT_FAX.as_bytes();
#[cfg(test)]
#[allow(unused)]
const B_FILTER_DCT_DECODE: &[u8] = FILTER_DCT_DECODE.as_bytes();
#[cfg(test)]
#[allow(unused)]
const B_FILTER_ASCII85_DECODE: &[u8] = FILTER_ASCII85_DECODE.as_bytes();
#[cfg(test)]
#[allow(unused)]
const B_FILTER_RUN_LENGTH_DECODE: &[u8] = FILTER_RUN_LENGTH_DECODE.as_bytes();
#[cfg(test)]
#[allow(unused)]
const B_FILTER_JPX_DECODE: &[u8] = FILTER_JPX_DECODE.as_bytes();

#[derive(Clone, PartialEq, Debug)]
pub struct Stream<'a>(Dictionary<'a>, &'a [u8]); // NOTE: buf end at the file end

/// error!() log if r is error, returns `Err<ObjectValueError::FilterDecodeError>`
fn handle_filter_error<V, E: Display>(
    r: Result<V, E>,
    filter_name: &str,
) -> Result<V, ObjectValueError> {
    r.map_err(|err| {
        error!("Failed to decode stream using {}: {}", filter_name, &err);
        ObjectValueError::FilterDecodeError
    })
}

fn decode_flate(buf: &[u8], params: Option<&Dictionary>) -> Result<Vec<u8>, ObjectValueError> {
    assert!(params.is_none(), "TODO: handle params of FlateDecode");

    use flate2::bufread::{DeflateDecoder, ZlibDecoder};
    use std::io::Read;

    let mut output = Vec::with_capacity(buf.len() * 2);
    let mut decoder = ZlibDecoder::new(buf);
    handle_filter_error(
        decoder
            .read_to_end(&mut output)
            .or_else(|_| DeflateDecoder::new(buf).read_to_end(&mut output)),
        FILTER_FLATE_DECODE,
    )?;

    Ok(output)
}

fn decode_dct<'a>(
    buf: Cow<'a, [u8]>,
    params: Option<&Dictionary>,
    image_to_raw: bool,
) -> Result<FilterDecodedData<'a>, ObjectValueError> {
    assert!(
        params.is_none(),
        "TODO: handle params of {}",
        FILTER_DCT_DECODE
    );

    if image_to_raw {
        return Ok(FilterDecodedData::RawImage(RawImage {
            format: ImageFormat::Jpeg,
            data: buf,
        }));
    }

    use image::{load_from_memory_with_format, ImageFormat as ImgImageFormat};
    handle_filter_error(
        load_from_memory_with_format(buf.borrow(), ImgImageFormat::Jpeg),
        FILTER_DCT_DECODE,
    )
    .map(FilterDecodedData::Image)
}

fn decode_jpx<'a>(
    buf: Cow<'a, [u8]>,
    params: Option<&Dictionary>,
    image_to_raw: bool,
) -> Result<FilterDecodedData<'a>, ObjectValueError> {
    assert!(
        params.is_none(),
        "TODO: handle params of {}",
        FILTER_JPX_DECODE
    );

    if image_to_raw {
        return Ok(FilterDecodedData::RawImage(RawImage {
            format: ImageFormat::Jpeg2k,
            data: buf,
        }));
    }

    use jpeg2k::Image;
    let img = handle_filter_error(Image::from_bytes(buf.borrow()), FILTER_JPX_DECODE)?;
    let img = handle_filter_error((&img).try_into(), FILTER_JPX_DECODE)?;
    Ok(FilterDecodedData::Image(img))
}

#[derive(Clone, Copy, PartialEq, Eq, Debug, strum::EnumString)]
enum ColorSpace {
    DeviceGray,
    DeviceRGB,
    DeviceCMYK,
    CalGray,
}

#[pdf_object((Some("XObject"), "Image"))]
trait ImageDictTrait {
    fn width(&self) -> u32;
    fn height(&self) -> u32;
    fn bits_per_component(&self) -> Option<u8>;
    #[from_name_str]
    fn color_space(&self) -> Option<ColorSpace>;
}

struct CCITTFaxDecodeParams<'a: 'b, 'b>(&'b Dictionary<'a>);

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum CCITTFGroup {
    #[allow(dead_code)]
    Group3_1D,
    #[allow(dead_code)]
    Group3_2D(i32),
    Group4,
}

impl<'a: 'b, 'b> From<&CCITTFaxDecodeParams<'a, 'b>> for Flags {
    fn from(params: &CCITTFaxDecodeParams<'a, 'b>) -> Self {
        assert!(!params.end_of_line());
        assert!(params.end_of_block());
        assert_eq!(0, params.damaged_rows_before_error());

        Flags {
            encoded_byte_align: params.encoded_byte_align(),
            inverse_black_white: params.black_is1(),
        }
    }
}

pub enum FilterDecodedData<'a> {
    Bytes(Cow<'a, [u8]>),
    Image(DynamicImage),
    RawImage(RawImage<'a>),
}

impl<'a> FilterDecodedData<'a> {
    fn bytes(bytes: Vec<u8>) -> Self {
        Self::Bytes(Cow::Owned(bytes))
    }

    /// Return [[ObjectValueError::StreamIsNotBytes]] if stream is not [[Self::Bytes]]
    fn into_bytes(self) -> Result<Cow<'a, [u8]>, ObjectValueError> {
        match self {
            Self::Bytes(bytes) => Ok(bytes),
            _ => Err(ObjectValueError::StreamIsNotBytes),
        }
    }

    /// Convert to bytes, for Image and RawImage returns image bytes.
    pub fn as_bytes(&self) -> &[u8] {
        match self {
            Self::Bytes(bytes) => bytes,
            Self::Image(img) => img.as_bytes(),
            Self::RawImage(img) => &img.data[..],
        }
    }
}

impl<'a: 'b, 'b> CCITTFaxDecodeParams<'a, 'b> {
    pub fn k(&self) -> CCITTFGroup {
        match self.0.get_int("K", 0).unwrap() {
            0 => CCITTFGroup::Group3_1D,
            k @ 1.. => CCITTFGroup::Group3_2D(k),
            ..=-1 => CCITTFGroup::Group4,
        }
    }

    pub fn end_of_line(&self) -> bool {
        self.0.get_bool("EndOfLine", false).unwrap()
    }

    pub fn encoded_byte_align(&self) -> bool {
        self.0.get_bool("EncodedByteAlign", false).unwrap()
    }

    pub fn columns(&self) -> u16 {
        self.0.get_int("Columns", 1728).unwrap() as u16
    }

    pub fn rows(&self) -> u16 {
        self.0.get_int("Rows", 0).unwrap() as u16
    }

    pub fn end_of_block(&self) -> bool {
        self.0.get_bool("EndOfBlock", true).unwrap()
    }

    pub fn black_is1(&self) -> bool {
        self.0.get_bool("BlackIs1", false).unwrap()
    }

    pub fn damaged_rows_before_error(&self) -> i32 {
        self.0.get_int("DamagedRowsBeforeError", 0).unwrap()
    }
}

fn decode_ascii85(
    buf: &[u8],
    params: Option<&Dictionary<'_>>,
) -> Result<Vec<u8>, ObjectValueError> {
    assert!(params.is_none());
    use crate::ascii85::decode;
    handle_filter_error(decode(buf), FILTER_ASCII85_DECODE)
}

fn decode_run_length(buf: &[u8], params: Option<&Dictionary<'_>>) -> Vec<u8> {
    assert!(params.is_none());
    use crate::run_length::decode;
    decode(buf)
}

fn decode_ccitt<'a: 'b, 'b>(
    input: &[u8],
    params: Option<&'b Dictionary<'a>>,
) -> Result<Vec<u8>, ObjectValueError> {
    {
        let params = params;
        use crate::ccitt::decode;

        let empty_params = Lazy::new(Dictionary::new);
        let params = CCITTFaxDecodeParams(params.unwrap_or_else(|| Lazy::force(&empty_params)));
        assert_eq!(params.k(), CCITTFGroup::Group4);
        let image = handle_filter_error(
            decode(
                input,
                params.columns(),
                Some(params.rows() as usize),
                (&params).into(),
            ),
            FILTER_CCITT_FAX,
        )?;
        Ok((image, (params.columns() as u32, params.rows() as u32)))
    }
    .map(|(buf, _meta)| buf)
}

fn filter<'a: 'b, 'b>(
    buf: Cow<'a, [u8]>,
    filter_name: &str,
    params: Option<&'b Dictionary<'a>>,
    image_to_raw: bool,
) -> Result<FilterDecodedData<'a>, ObjectValueError> {
    match filter_name {
        FILTER_FLATE_DECODE => decode_flate(&buf, params).map(FilterDecodedData::bytes),
        FILTER_DCT_DECODE => decode_dct(buf, params, image_to_raw),
        FILTER_CCITT_FAX => decode_ccitt(&buf, params).map(FilterDecodedData::bytes),
        FILTER_ASCII85_DECODE => decode_ascii85(&buf, params).map(FilterDecodedData::bytes),
        FILTER_RUN_LENGTH_DECODE => Ok(FilterDecodedData::bytes(decode_run_length(&buf, params))),
        FILTER_JPX_DECODE => decode_jpx(buf, params, image_to_raw),
        _ => {
            error!("Unknown filter: {}", filter_name);
            Err(ObjectValueError::UnknownFilter)
        }
    }
}

pub enum ImageFormat {
    Jpeg,
    Jpeg2k,
    Png,
}

pub struct RawImage<'a> {
    pub format: ImageFormat,
    pub data: Cow<'a, [u8]>,
}

impl<'a> Stream<'a> {
    pub fn new(dict: Dictionary<'a>, data: &'a [u8]) -> Self {
        Self(dict, data)
    }

    /// Decode stream data using filter and parameters in stream dictionary.
    /// `image_to_raw` if the stream is image, convert to RawImage.
    pub fn decode(
        &self,
        resolver: &ObjectResolver<'a>,
        image_to_raw: bool,
    ) -> Result<FilterDecodedData<'a>, ObjectValueError> {
        let len = resolver
            .resolve_container_value(&self.0, "Length")?
            .1
            .as_int()?;

        #[cfg(debug_assertions)]
        {
            let end_stream = &self.1[len as usize..];
            let rv: ParseResult<_> =
                ws_prefixed(nom::bytes::complete::tag(b"endstream"))(end_stream);
            if rv.is_err() {
                panic!("{:#?}", self.1);
            }
        }

        let mut decoded = FilterDecodedData::Bytes(Cow::Borrowed(&self.1[0..len as usize]));
        for (filter_name, params) in self.iter_filter()? {
            decoded = filter(decoded.into_bytes()?, filter_name, params, image_to_raw)?;
        }

        let img_dict = ImageDict::from(&self.0, resolver)?;
        let Some(img_dict) = img_dict else {
            return Ok(decoded);
        };

        let FilterDecodedData::Bytes(data) = decoded else {
            return Ok(decoded);
        };

        if image_to_raw {
            match (
                img_dict.color_space(),
                img_dict.bits_per_component().unwrap(),
            ) {
                (Some(ColorSpace::DeviceGray), 1) => {
                    use png::{BitDepth, ColorType, Encoder};
                    let mut bytes = Vec::new();
                    let mut encoder = Encoder::new(&mut bytes, img_dict.width(), img_dict.height());
                    encoder.set_color(ColorType::Grayscale);
                    encoder.set_depth(BitDepth::One);
                    let mut writer = encoder.write_header().unwrap();
                    writer.write_image_data(data.borrow()).unwrap();
                    drop(writer);
                    Ok(FilterDecodedData::RawImage(RawImage {
                        format: ImageFormat::Png,
                        data: Cow::Owned(bytes),
                    }))
                }
                (Some(ColorSpace::DeviceGray), 8) => {
                    use png::{BitDepth, ColorType, Encoder};
                    let mut bytes = Vec::new();
                    let mut encoder = Encoder::new(&mut bytes, img_dict.width(), img_dict.height());
                    encoder.set_color(ColorType::Grayscale);
                    encoder.set_depth(BitDepth::Eight);
                    let mut writer = encoder.write_header().unwrap();
                    writer.write_image_data(data.borrow()).unwrap();
                    drop(writer);
                    Ok(FilterDecodedData::RawImage(RawImage {
                        format: ImageFormat::Png,
                        data: Cow::Owned(bytes),
                    }))
                }
                (Some(ColorSpace::DeviceRGB), 8) => {
                    use png::{BitDepth, ColorType, Encoder};
                    let mut bytes = Vec::new();
                    let mut encoder = Encoder::new(&mut bytes, img_dict.width(), img_dict.height());
                    encoder.set_color(ColorType::Rgb);
                    encoder.set_depth(BitDepth::Eight);
                    let mut writer = encoder.write_header().unwrap();
                    writer.write_image_data(data.borrow()).unwrap();
                    drop(writer);
                    Ok(FilterDecodedData::RawImage(RawImage {
                        format: ImageFormat::Png,
                        data: Cow::Owned(bytes),
                    }))
                }
                _ => todo!(
                    "unsupported interoperate decoded stream data as raw image: {:?} {}",
                    img_dict.color_space(),
                    img_dict.bits_per_component().unwrap()
                ),
            }
        } else {
            match (
                img_dict.color_space(),
                img_dict.bits_per_component().unwrap(),
            ) {
                (_, 1) => {
                    use bitstream_io::read::BitRead;

                    let mut img = GrayImage::new(img_dict.width(), img_dict.height());
                    let mut r = BitReader::<_, BigEndian>::new(data.borrow() as &[u8]);
                    for y in 0..img_dict.height() {
                        for x in 0..img_dict.width() {
                            img.put_pixel(
                                x,
                                y,
                                Luma([if r.read_bit().unwrap() { 255u8 } else { 0 }]),
                            );
                        }
                    }
                    Ok(FilterDecodedData::Image(DynamicImage::ImageLuma8(img)))
                }
                (Some(ColorSpace::DeviceGray), 8) => {
                    let img =
                        GrayImage::from_raw(img_dict.width(), img_dict.height(), data.into_owned())
                            .unwrap();
                    Ok(FilterDecodedData::Image(DynamicImage::ImageLuma8(img)))
                }
                (Some(ColorSpace::DeviceRGB), 8) => {
                    let img =
                        RgbImage::from_raw(img_dict.width(), img_dict.height(), data.into_owned())
                            .unwrap();
                    Ok(FilterDecodedData::Image(DynamicImage::ImageRgb8(img)))
                }
                _ => todo!(
                    "unsupported interoperate decoded stream data as image: {:?} {}",
                    img_dict.color_space(),
                    img_dict.bits_per_component().unwrap()
                ),
            }
        }
    }

    fn iter_filter(
        &self,
    ) -> Result<impl Iterator<Item = (&str, Option<&Dictionary<'a>>)>, ObjectValueError> {
        if self.0.contains_key(&Name::borrowed(KEY_FFILTER)) {
            return Err(ObjectValueError::ExternalStreamNotSupported);
        }

        let filters = self.0.get(&Name::borrowed(KEY_FILTER)).map_or_else(
            || Ok(vec![]),
            |v| match v {
                Object::Array(vals) => vals
                    .iter()
                    .map(|v| v.as_name().map_err(|_| ObjectValueError::UnexpectedType))
                    .collect(),
                Object::Name(n) => Ok(vec![from_utf8(n.0.borrow()).unwrap()]),
                _ => Err(ObjectValueError::UnexpectedType),
            },
        )?;
        let params = self.0.get(&Name::borrowed(KEY_FILTER_PARAMS)).map_or_else(
            || Ok(vec![]),
            |v| match v {
                Object::Null => Ok(vec![]),
                Object::Array(vals) => vals
                    .iter()
                    .map(|v| match v {
                        Object::Null => Ok(None),
                        Object::Dictionary(dict) => Ok(Some(dict)),
                        _ => Err(ObjectValueError::UnexpectedType),
                    })
                    .collect(),
                Object::Dictionary(dict) => Ok(vec![Some(dict)]),
                _ => Err(ObjectValueError::UnexpectedType),
            },
        )?;
        Ok(filters
            .into_iter()
            .zip(params.into_iter().chain(repeat(None))))
    }
}

#[cfg(test)]
mod tests;
