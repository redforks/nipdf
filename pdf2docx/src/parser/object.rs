use std::borrow::Cow;

use nom::{
    branch::alt,
    bytes::{
        complete::{escaped, is_not, tag, take_till, take_until, take_while},
        streaming::take,
    },
    character::{
        complete::{anychar, crlf, multispace1, u16, u32},
        is_hex_digit,
    },
    combinator::{map, not, opt, peek, recognize, value},
    multi::{many0, many0_count},
    number::complete::float,
    sequence::{delimited, preceded, separated_pair, terminated, tuple},
};
use num::cast;

use crate::object::{
    Array, Dictionary, HexString, IndirectObject, LiteralString, Name, Object, ObjectValueError,
    Reference, Stream,
};

use super::{whitespace_or_comment, ws, ws_prefixed, ws_terminated, ParseError, ParseResult};

/// Unwrap the result of nom parser to a *normal* result.
pub fn unwrap_parse_result<'a, T: 'a>(obj: ParseResult<'a, T>) -> Result<T, ParseError<'a>> {
    match obj {
        Ok((_, obj)) => Ok(obj),
        Err(nom::Err::Incomplete(_)) => unreachable!(),
        Err(nom::Err::Error(e)) | Err(nom::Err::Failure(e)) => Err(e),
    }
}

pub fn parse_object(buf: &[u8]) -> ParseResult<Object> {
    let null = value(Object::Null, tag(b"null"));
    let true_parser = value(Object::Bool(true), tag(b"true"));
    let false_parser = value(Object::Bool(false), tag(b"false"));
    let number_parser = map(float, |v| {
        if let Some(i) = cast(v) {
            if v.fract() == 0.0 {
                Object::Integer(i)
            } else {
                Object::Number(v)
            }
        } else {
            Object::Number(v)
        }
    });

    fn parse_quoted_string(input: &[u8]) -> ParseResult<&[u8]> {
        let esc = escaped(is_not("\\()"), '\\', anychar);
        let inner_parser = alt((esc, parse_quoted_string));
        let mut parser = recognize(delimited(tag(b"("), many0_count(inner_parser), tag(b")")));
        parser(input)
    }
    let parse_quoted_string = map(parse_quoted_string, |s| {
        Object::LiteralString(LiteralString::new(s))
    });
    let parse_hex_string = map(
        recognize(delimited(
            tag(b"<"),
            take_while(|c| is_hex_digit(c) || c.is_ascii_whitespace()),
            tag(b">"),
        )),
        |buf| Object::HexString(HexString::new(buf)),
    );

    alt((
        map(parse_name, Object::Name),
        parse_quoted_string,
        map(parse_dict, Object::Dictionary),
        map(parse_array, Object::Array),
        parse_hex_string,
        null,
        true_parser,
        false_parser,
        map(parse_reference, Object::Reference),
        number_parser,
    ))(buf)
}

/// Return `Err(ObjectValueError::InvalidNameForma)` if the name is not a valid PDF name encoding,
/// not two hex char after `#`.
fn normalize_name(buf: &[u8]) -> Result<Cow<[u8]>, ObjectValueError> {
    fn next_hex_char(iter: &mut impl Iterator<Item = u8>) -> Option<u8> {
        let mut result = 0;
        for _ in 0..2 {
            if let Some(c) = iter.next() {
                result <<= 4;
                result |= match c {
                    b'0'..=b'9' => c - b'0',
                    b'a'..=b'f' => c - b'a' + 10,
                    b'A'..=b'F' => c - b'A' + 10,
                    _ => return None,
                };
            } else {
                return None;
            }
        }
        Some(result)
    }

    let s = &buf[1..];
    if s.iter().copied().any(|b| b == b'#') {
        let mut result = Vec::with_capacity(s.len());
        let mut iter = s.iter().copied();
        while let Some(next) = iter.next() {
            if next == b'#' {
                if let Some(c) = next_hex_char(&mut iter) {
                    result.push(c);
                } else {
                    return Err(ObjectValueError::InvalidNameFormat);
                }
            } else {
                result.push(next);
            }
        }
        Ok(Cow::Owned(result))
    } else {
        Ok(Cow::Borrowed(s))
    }
}

fn parse_name(input: &[u8]) -> ParseResult<Name> {
    let (input, buf) = recognize(preceded(
        tag(b"/"),
        take_till(|c: u8| {
            c.is_ascii_whitespace()
                || c == b'['
                || c == b'<'
                || c == b'('
                || c == b'/'
                || c == b'>'
                || c == b']'
        }),
    ))(input)?;
    let name = normalize_name(buf)
        .map_err(|_| nom::Err::Error(ParseError::InvalidNameFormat))
        .map(Name)?;
    Ok((input, name))
}

pub fn parse_array(input: &[u8]) -> ParseResult<Array> {
    delimited(
        ws(tag(b"[")),
        many0(ws(parse_object)),
        ws_terminated(tag(b"]")),
    )(input)
}

pub fn parse_dict(input: &[u8]) -> ParseResult<Dictionary> {
    map(
        delimited(
            ws(tag(b"<<".as_slice())),
            many0(tuple((parse_name, ws(parse_object)))),
            ws_terminated(tag(b">>")),
        ),
        |v| v.into_iter().collect(),
    )(input)
}

pub fn parse_stream_content(input: &[u8], stream_len: u32) -> ParseResult<&[u8]> {
    delimited(
        tuple((ws_prefixed(tag(b"stream")), alt((crlf, tag(b"\n"))))),
        take(stream_len),
        ws(tag(b"endstream")),
    )(input)
}

fn parse_object_and_stream(input: &[u8]) -> ParseResult<Object> {
    let (input, o) = parse_object(input)?;
    let (input, buf) = match o {
        Object::Dictionary(ref d) => {
            let Some(len) = d.get(&Name::borrowed(b"Length")) else {
                return Ok((input, o));
            };
            match len {
                Object::Integer(len) => opt(delimited(
                    tuple((ws_prefixed(tag(b"stream")), alt((crlf, tag(b"\n"))))),
                    take(*len as u32),
                    ws(tag(b"endstream")),
                ))(input)?,
                Object::Reference(_) => {
                    let (input, inner) = peek(opt(ws_prefixed(tag(b"stream"))))(input)?;
                    return match inner {
                        None => Ok((input, o)),
                        Some(_) => {
                            let (input, _) =
                                take_until::<&[u8], &[u8], ParseError>(b"endobj")(input)?;
                            Ok((input, Object::LaterResolveStream(d.clone())))
                        }
                    };
                }
                _ => return Ok((input, o)),
            }
        }
        _ => return Ok((input, o)),
    };
    match buf {
        Some(buf) => match o {
            Object::Dictionary(d) => Ok((input, Object::Stream(Stream(d, buf)))),
            _ => unreachable!(),
        },
        None => Ok((input, o)),
    }
}

pub fn parse_indirected_object(input: &[u8]) -> ParseResult<IndirectObject> {
    let (input, (id, gen)) = separated_pair(u32, multispace1, u16)(input)?;
    let (input, obj) =
        delimited(ws(tag(b"obj")), parse_object_and_stream, ws(tag(b"endobj")))(input)?;
    Ok((input, IndirectObject::new(id, gen, obj)))
}

fn parse_reference(input: &[u8]) -> ParseResult<Reference> {
    let (input, (id, gen)) = terminated(
        separated_pair(u32, multispace1, u16),
        // `not(peek(tag("G")))` to detect `RG` graphics operation,
        // such as `0 1 0 RG` is a valid graphics operation,
        // if not check `not(peek(tag("G")))`, the sequence will parsed to
        // Integer(0), Reference(1 0)
        delimited(whitespace_or_comment, tag(b"R"), not(peek(tag("G")))),
    )(input)?;
    Ok((input, Reference::new(id, gen)))
}

#[cfg(test)]
mod tests;
