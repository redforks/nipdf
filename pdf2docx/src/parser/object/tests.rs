

use super::*;
use crate::object::HexString;
use test_case::test_case;

#[test_case(Object::Null, "null"; "null")]
#[test_case(true, "true")]
#[test_case(false, "false")]
#[test_case(123, "123"; "integer")]
#[test_case(-123, "-123"; "negative integer")]
#[test_case(123.12, "123.12"; "number")]
#[test_case(-123.12, "-123.12"; "negative number")]
#[test_case(LiteralString::new(b"()"), "()"; "empty literal string")]
#[test_case(LiteralString::new(b"(5\\()"), "(5\\()"; "escaped )")]
#[test_case(LiteralString::new(b"(5\\\\)"), "(5\\\\)"; "escaped back slash")]
#[test_case(LiteralString::new(b"(a(foo))"), "(a(foo))"; "nested quoted string")]
#[test_case(LiteralString::new(b"(a
b)"), "(a
b)"; "literal string contains new line")]
#[test_case(LiteralString::new(b"(*!&}^%)"), "(*!&}^%)"; "literal string contains special characters")]
#[test_case(LiteralString::new(b"(\\)\\()"), "(\\)\\()"; "literal string contains escape")]
#[test_case(LiteralString::new(b"(\\333\\n)"), "(\\333\\n)")]
#[test_case(HexString::new(b"<>"), "<>")]
#[test_case(HexString::new(b"<12A>"), "<12A>")]
#[test_case(HexString::new(b"<12 A\t3>"), "<12 A\t3>"; "contains whitespace")]
#[test_case(Name::borrowed(b""), "/"; "empty name")]
#[test_case(Name::borrowed(b"foo"), "/foo"; "name")]
fn test_parse_simple_objects(exp: impl Into<Object<'static>>, buf: &'static str) {
    let o = parse_object(buf.as_bytes()).unwrap();
    assert_eq!((b"".as_slice(), exp.into()), o);
}

#[test_case(vec![], "[]"; "empty array")]
#[test_case(vec![], "[ \t]"; "empty array 2")]
#[test_case(vec![Object::Null], "[null]"; "array with null")]
#[test_case(vec![Object::Array(vec![Object::Null])], "[[null]]"; "nested array with null")]
#[test_case(vec![Name::borrowed(b"foo").into()], "[/foo]"; "name value")]
fn test_parse_array(exp: Vec<Object<'static>>, buf: &'static str) {
    assert_eq!((b"".as_slice(), exp), parse_array(buf.as_bytes()).unwrap());
}

#[test_case(b"<< >>", "empty dict")]
#[test_case(b"<<>>", "empty dict 2")]
#[test_case(b"<< /Type /Catalog >>", "dict with one entry")]
#[test_case(b"<</Inner<<>>>>", "nested")]
#[test_case(b"<</id[]>>", "empty array")]
#[test_case(b"<</id()>>", "string value")]
#[test_case(b"<</id/Value>>", "name value")]
#[test_case(b"<</id/>>", "empty name value")]
#[test_case(b"<<//id>>", "empty name key")]
#[test_case(b"<</id<<>>>>", "nested empty dict")]
fn test_parse_dict(buf: impl AsRef<[u8]>, name: &str) {
    insta::assert_debug_snapshot!(name, parse_dict(buf.as_ref()).unwrap());
}

#[test_case(
    b"1 0 obj
null
endobj",
    "null"
)]
fn test_parse_indirected_object(buf: impl AsRef<[u8]>, name: &str) {
    insta::assert_debug_snapshot!(name, parse_indirected_object(buf.as_ref()).unwrap());
}

#[test_case(b"1 0 R", "simple")]
fn test_parse_reference(buf: impl AsRef<[u8]>, name: &str) {
    insta::assert_debug_snapshot!(name, parse_reference(buf.as_ref()).unwrap());
}

#[test_case(b"foo", b"/foo")]
#[test_case(b"a#b", b"/a#23b")]
#[test_case(b"Ab", b"/#41#62")]
fn name_normalize(exp: impl AsRef<[u8]>, name: impl AsRef<[u8]>) {
    assert_eq!(normalize_name(name.as_ref()).unwrap(), exp.as_ref());
}
