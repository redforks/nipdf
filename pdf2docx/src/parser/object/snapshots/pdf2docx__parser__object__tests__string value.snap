---
source: pdf2docx/src/parser/object/tests.rs
expression: parse_dict(buf.as_ref()).unwrap()
---
(
    [],
    Dictionary(
        {
            Name(
                [
                    105,
                    100,
                ],
            ): LiteralString(
                LiteralString(
                    [
                        40,
                        41,
                    ],
                ),
            ),
        },
    ),
)
